src/index_notation/transformations.cpp

```diff
@@ -1487,6 +1492,293 @@ ir::Stmt GEMM::replaceValidStmt(IndexStmt stmt,
+  auto bStrides = ir::Var::make("bStrides", Datatype("std::array<tblis::stride_type, " + util::toString(bTensorVar.getOrder()) + ">"));
+  auto cStrides = ir::Var::make("cStrides", Datatype("std::array<tblis::stride_type, " + util::toString(cTensorVar.getOrder()) + ">"));
+
+  results.push_back(ir::VarDecl::make(aStrides, ir::Symbol::make("{}")));
+  for (int i = 0; i < aTensorVar.getOrder(); ++i) {
+    results.push_back(ir::Store::make(aStrides, i, getStride(aAccess, i)));
+  }
+
+  results.push_back(ir::VarDecl::make(bStrides, ir::Symbol::make("{}")));
+  for (int i = 0; i < bTensorVar.getOrder(); ++i) {
+    results.push_back(ir::Store::make(bStrides, i, getStride(bAccess, i)));
+  }
+
+  results.push_back(ir::VarDecl::make(cStrides, ir::Symbol::make("{}")));
+  for (int i = 0; i < cTensorVar.getOrder(); ++i) {
+    results.push_back(ir::Store::make(cStrides, i, getStride(cAccess, i)));
+  }
+
+  // Create TBLIS tensors.
+  Datatype TblisTensor("tblis::tblis_tensor");
+  auto aTensor = ir::Var::make("aTensor", TblisTensor);
+  auto bTensor = ir::Var::make("bTensor", TblisTensor);
+  auto cTensor = ir::Var::make("cTensor", TblisTensor);
+
+  std::string initTensorFunc;
+  if (type.getKind() == Float32) {
+    initTensorFunc = "tblis::tblis_init_tensor_s";
+  } else if (type.getKind() == Float64) {
+    initTensorFunc = "tblis::tblis_init_tensor_d";
+  } else {
+    taco_iassert(false);
+  }
+
+  std::vector<ir::Expr> aArgs = {
+    ir::Call::make("&", {aTensor}, Auto),
+    aTensorVar.getOrder(), // ndim
+    ir::MethodCall::make(aShape, "data", {}, false /* deref */, Auto), // shape
+    getPtr(aAccess, aDomain), // data
+    ir::MethodCall::make(aStrides, "data", {}, false /* deref */, Auto), // strides
+  };
+  results.push_back(ir::VarDecl::make(aTensor, ir::Symbol::make("{}")));
+  results.push_back(ir::SideEffect::make(ir::Call::make(initTensorFunc, aArgs, Auto)));
+
+  std::vector<ir::Expr> bArgs = {
+    ir::Call::make("&", {bTensor}, Auto),
+    bTensorVar.getOrder(),
+    ir::MethodCall::make(bShape, "data", {}, false /* deref */, Auto),
+    ir::Call::make(unconst_cast, {getPtr(bAccess, aDomain)}, Auto),
+    ir::MethodCall::make(bStrides, "data", {}, false /* deref */, Auto),
+  };
+  results.push_back(ir::VarDecl::make(bTensor, ir::Symbol::make("{}")));
+  results.push_back(ir::SideEffect::make(ir::Call::make(initTensorFunc, bArgs, Auto)));
+
+  std::vector<ir::Expr> cArgs = {
+    ir::Call::make("&", {cTensor}, Auto),
+    cTensorVar.getOrder(), // ndim
+    ir::MethodCall::make(cShape, "data", {}, false /* deref */, Auto),
+    ir::Call::make(unconst_cast, {getPtr(cAccess, aDomain)}, Auto),
+    ir::MethodCall::make(cStrides, "data", {}, false /* deref */, Auto),
+  };
+  results.push_back(ir::VarDecl::make(cTensor, ir::Symbol::make("{}")));
+  results.push_back(ir::SideEffect::make(ir::Call::make(initTensorFunc, cArgs, Auto)));
+
+  std::vector<ir::Expr> args = {
+      ir::Symbol::make("tblis_single"),
```

Comment:
I think that tblis_single is the wrong thing here still. Looking at the docs, `tblis_single` is for single-threaded execution mode, which we don't want. We want to utilize the available openmp threads. I believe that this is done by passing `nullptr`, ala https://github.com/nv-legate/cunumeric/blob/branch-22.05/src/cunumeric/matrix/contract_omp.cc.

Reply:
Hmm, I tried using nullptr but the resultant binary appears to hang. When I tried to do the same computation in a new file without any Legion stuff though, it works. We can maybe check this out tomorrow.

Reply:
Yeah, that is probably due to something Legion related then. If you're interested, try running `ldd <binaryname>` to see if `libgomp` has been linked into your binary. If it has, then that's the problem: Legion provides a custom implementation of OpenMP that must be linked against. Sometimes the system openmp can creep in depending on what compilation flags are specified during building of the library or the binary.

cunumeric builds tblis in this way, maybe it's slightly different than how you did it? https://github.com/nv-legate/cunumeric/blob/a42cadccf6efcb94aa30c9c540a714a9b837d484/install.py#L207

Reply:
Yeah, it sounds like libgomp is the problem. libtblis.so and libtci.so both link to libgomp.so. I see that `legion/legion/cmake-install/lib/librealm.a` has a bunch of `gomp_*` symbols defined. Is that the special library you are talking about?

cuNumeric specifies its own length/stride/label type, but the default should work fine for us.

Reply:
Also, I see that libopenblas.so.0 also links to libgomp.so, but it doesn't seem to be an issue?

Reply:
Yes, Realm (a subsystem of Legion) provides the OpenMP library implementation. I would imagine that `libtblis` and `libtci` should have references to openmp symbols but shouldn't pull in the library themselves.

Sometimes this is caused by linking the tblis library after Legion (the linker resolves symbols in a certain order).

Reply:
> Also, I see that libopenblas.so.0 also links to libgomp.so, but it doesn't seem to be an issue?

Yeah, this is actually OK i think. I checked with a cunumeric developer, and the output of ldd on their libtblis has libgomp. But he said that basically we need to ensure that librealm is loaded before libtblis so that symbol resolution of openmp related symbols resolve to those in librealm.

However, he says that this shouldn't cause a hang, if this has something to do with Legion's OpenMP module. It's possible however, that this could run extremely slow if the openmp calls aren't getting routed through Legion. One thing to confirm is whether or not it is hanging or just running slowly.

Reply:
Fixed by calling `tblis_set_num_threads` before invoking tblis::mult, in the case where OpenMP is enabled.

https://github.com/rohany/taco/blob/5a9ff27eb47a8cdf821e14cc4260405de7ade404/legion/include/leaf_kernels.h#L144-L151